### PR TITLE
Replace DUMMY_TRANSACTION

### PR DESCRIPTION
**Pull request by hououou**
Refer to #3062 

Remove dummy transactions and propagate client context.

### CODE REVIEW

File: src/binder/bind/bind_updating_clause.cpp [hunk 1]
```diff
@@ -96,7 +96,8 @@ std::vector<BoundInsertInfo> Binder::bindInsertInfos(
     const QueryGraphCollection& queryGraphCollection, const expression_set& nodeRelScope_) {
     auto nodeRelScope = nodeRelScope_;
     std::vector<BoundInsertInfo> result;
-    auto analyzer = QueryGraphLabelAnalyzer(*clientContext->getCatalog());
+    auto analyzer = QueryGraphLabelAnalyzer(*clientContext->getCatalog(), *clientContext);
+    // auto analyzer = QueryGraphLabelAnalyzer(catalog, *clientContext);
```
Comment (andyfengHKU) [inline src/binder/bind/bind_updating_clause.cpp:100]: remove

File: src/include/binder/bound_statement_rewriter.h [hunk 2]
```diff
@@ -9,7 +9,8 @@ namespace binder {
 // Perform semantic rewrite over bound statement.
 class BoundStatementRewriter {
 public:
-    static void rewrite(BoundStatement& boundStatement, const catalog::Catalog& catalog);
+    static void rewrite(BoundStatement& boundStatement, const catalog::Catalog& catalog,
```
Comment (andyfengHKU) [inline src/include/binder/bound_statement_rewriter.h:12]: Keep track that you next PR should remove catalog from this function

File: src/include/binder/query/query_graph_label_analyzer.h [hunk 3]
```diff
@@ -8,7 +8,9 @@ namespace binder {
 
 class QueryGraphLabelAnalyzer {
 public:
-    explicit QueryGraphLabelAnalyzer(const catalog::Catalog& catalog) : catalog{catalog} {}
+    explicit QueryGraphLabelAnalyzer(
```
Comment (andyfengHKU) [inline src/include/binder/query/query_graph_label_analyzer.h:11]: Ditto.

File: src/include/binder/rewriter/match_clause_pattern_label_rewriter.h [hunk 4]
```diff
@@ -8,7 +8,9 @@ namespace binder {
 
 class MatchClausePatternLabelRewriter : public BoundStatementVisitor {
 public:
-    explicit MatchClausePatternLabelRewriter(const catalog::Catalog& catalog) : analyzer{catalog} {}
+    explicit MatchClausePatternLabelRewriter(
```
Comment (andyfengHKU) [inline src/include/binder/rewriter/match_clause_pattern_label_rewriter.h:11]: Ditto

File: src/include/planner/planner.h [hunk 5]
```diff
@@ -32,7 +32,8 @@ struct LogicalSetPropertyInfo;
 
 class Planner {
 public:
-    Planner(catalog::Catalog* catalog, storage::StorageManager* storageManager);
+    Planner(catalog::Catalog* catalog, storage::StorageManager* storageManager,
```
Comment (andyfengHKU) [inline src/include/planner/planner.h:35]: Remove catalog and storageManager in your next PR.

File: src/planner/join_order/cardinality_estimator.cpp [hunk 6]
```diff
@@ -10,25 +10,26 @@ using namespace kuzu::common;
 namespace kuzu {
 namespace planner {
 
-void CardinalityEstimator::initNodeIDDom(const QueryGraph& queryGraph) {
+void CardinalityEstimator::initNodeIDDom(
```
Comment (andyfengHKU) [inline src/planner/join_order/cardinality_estimator.cpp:13]: use namespace transaction and remove `transaction::`

File: src/planner/plan/append_extend.cpp [hunk 7]
```diff
@@ -22,7 +21,8 @@ namespace kuzu {
 namespace planner {
 
 static bool extendHasAtMostOneNbrGuarantee(RelExpression& rel, NodeExpression& boundNode,
-    ExtendDirection direction, const catalog::Catalog& catalog) {
+    ExtendDirection direction, const catalog::Catalog& catalog,
```
Comment (andyfengHKU) [inline src/planner/plan/append_extend.cpp:24]: remove catalog from this function in this PR.

File: src/planner/plan/append_extend.cpp [hunk 8]
```diff
@@ -34,16 +34,17 @@ static bool extendHasAtMostOneNbrGuarantee(RelExpression& rel, NodeExpression& b
...
-        catalog.getTableCatalogEntry(&DUMMY_READ_TRANSACTION, rel.getSingleTableID()));
+        catalog.getTableCatalogEntry(clientContext.getTx(), rel.getSingleTableID()));
     return relTableEntry->isSingleMultiplicity(relDirection);
 }
 
-static std::unordered_set<table_id_t> getBoundNodeTableIDSet(
-    const RelExpression& rel, ExtendDirection extendDirection, const catalog::Catalog& catalog) {
+static std::unordered_set<table_id_t> getBoundNodeTableIDSet(const RelExpression& rel,
+    ExtendDirection extendDirection, const catalog::Catalog& catalog,
+    const main::ClientContext& clientContext) {
```
Comment (andyfengHKU) [inline src/planner/plan/append_extend.cpp:43]: ditto

File: src/planner/plan/append_extend.cpp [hunk 9]
```diff
@@ -62,12 +63,13 @@ static std::unordered_set<table_id_t> getBoundNodeTableIDSet(
     return result;
 }
 
-static std::unordered_set<table_id_t> getNbrNodeTableIDSet(
-    const RelExpression& rel, ExtendDirection extendDirection, const catalog::Catalog& catalog) {
+static std::unordered_set<table_id_t> getNbrNodeTableIDSet(const RelExpression& rel,
+    ExtendDirection extendDirection, const catalog::Catalog& catalog,
+    const main::ClientContext& clientContext) {
```
Comment (andyfengHKU) [inline src/planner/plan/append_extend.cpp:68]: ditto

File: src/processor/map/map_delete.cpp [hunk 10]
```diff
@@ -13,15 +12,15 @@ namespace kuzu {
 namespace processor {
 
 static std::unique_ptr<NodeDeleteExecutor> getNodeDeleteExecutor(Catalog* catalog,
-    StorageManager& storageManager, LogicalDeleteNodeInfo* info, const Schema& inSchema) {
+    StorageManager& storageManager, LogicalDeleteNodeInfo* info, const Schema& inSchema,
+    main::ClientContext* clientContext) {
```
Comment (andyfengHKU) [inline src/processor/map/map_delete.cpp:16]: remove catalog and storageManager in this function

File: src/processor/map/map_extend.cpp [hunk 11]
```diff
@@ -32,11 +31,12 @@ static std::unique_ptr<ScanRelTableInfo> getRelTableScanInfo(TableCatalogEntry*
 
 static std::unique_ptr<RelTableCollectionScanner> populateRelTableCollectionScanner(
     table_id_t boundNodeTableID, const RelExpression& rel, ExtendDirection extendDirection,
-    const expression_vector& properties, StorageManager& storageManager, const Catalog& catalog) {
+    const expression_vector& properties, StorageManager& storageManager, const Catalog& catalog,
```
Comment (andyfengHKU) [inline src/processor/map/map_extend.cpp:34]: remove catalog and storageManager from this function

File: src/storage/stats/nodes_store_statistics.cpp [hunk 12]
```diff
@@ -22,8 +22,9 @@ offset_t NodesStoreStatsAndDeletedIDs::getMaxNodeOffset(
 std::map<table_id_t, offset_t> NodesStoreStatsAndDeletedIDs::getMaxNodeOffsetPerTable() const {
     std::map<table_id_t, offset_t> retVal;
     for (auto& tableIDStatistics : readOnlyVersion->tableStatisticPerTable) {
-        retVal[tableIDStatistics.first] =
-            getNodeStatisticsAndDeletedIDs(tableIDStatistics.first)->getMaxNodeOffset();
+        retVal[tableIDStatistics.first] = getNodeStatisticsAndDeletedIDs(
+            &transaction::DUMMY_READ_TRANSACTION, tableIDStatistics.first)
```
Comment (andyfengHKU) [inline src/storage/stats/nodes_store_statistics.cpp:26]: why we don't remove this DUMMY_READ_TRANSACTION?

File: src/storage/store/node_table.cpp [hunk 13]
```diff
@@ -135,7 +135,9 @@ void NodeTable::prepareCommit(Transaction* transaction, LocalTable* localTable)
     if (pkIndex) {
         pkIndex->prepareCommit();
     }
-    tableData->prepareLocalTableToCommit(transaction, localTable->getLocalTableData(0));
+    if (localTable) {
```
Comment (ray6080) [inline src/storage/store/node_table.cpp:138]: Why would we have `nullptr` here for `localTable`?

File: src/storage/storage_manager.cpp
Comment (ray6080) [inline src/storage/storage_manager.cpp:1]: Can you explain the changes around `deleteTables`? why do we erase the dropped table only until `rollbackInMemory`?

**Review by andyfengHKU — APPROVED** [review]

**Pull request merged**